### ISSUE TITLE
Is hook mode needed to uniquely identify reactions?

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#413]
After #410, we pass the slug of the hook mode that a reaction store relates to to the store object when it is constructed. We didn't add a way to retrieve the mode from the reaction store yet, because we weren't sure whether such a method should be part of the reaction store interface.

As I was working on this I realized that what we were doing in essence is making it so that a reaction store has to be identified based on two items: its slug and the hook mode it is associated with. It needs both of these to be able to prevent conflicts in how reaction stores with the same slug but for different modes are stored in the database. So really we weren't *doing* this just now, it is already this way, we just weren't doing it right before.

Anyway, this made me wonder whether we should also then be including the mode in the reaction GUID. Otherwise we can't uniquely identify the store to which the reaction belongs, which means we don't really have a GUID. But I'm not 100% sure about this. We need to explore a little more how the interplay between mode and context works, so that we have a good grasp on why this is (or isn't) needed. We already make the reaction store context a part of the reaction GUID, so why isn't that enough?

**Comment by JDGrimes** [comment]
Well, there are really three things here:

- hook mode
- context
- database contexts

Of course contexts within the site at run-time relate to contexts within the database. But the site at run-time is always *in* one particular context or other, while the database isn't exactly. The database describes those contexts "at rest", by its structure, etc.

**Comment by JDGrimes** [comment]
The reason I bring up the DB, by the way, is because that is where many reaction stores store the data. (All of the ones that we have right now, actually.) So there is a reaction store that uses options, and thus its reactions are stored in the context of that site in the database. The reaction store declares this via the `$context` property. That means that this reaction store can't work when we are in network mode, because we are outside of site context. However, several different modes beside just `standard` can theoretically exist in the site context. Because of this we can't just plop our data into an option name suffixed with the store slug. There might be another store with that slug registered for another mode, but using the same storage method. That would cause a conflict, which is why we also need to incorporate the mode that the store relates to into the option name.

**Comment by JDGrimes** [comment]
In other words, reaction stores have a GUID composed of:

- the context GUID
- the store slug
- the mode name

Which means that we need to use the mode as part of a reaction's GUID.

**Comment by JDGrimes** [comment]
The reason that this is so hard to get clear is that we don't really use multiple modes in a single context right now (`standard` -> `site`, `network` -> `network`). But I'm coming to grips with it.

**Comment by JDGrimes** [comment]
The question then is, how do we approach the fix? I'm wondering whether it would be better to introduce a `get_GUID()` method for the reaction store interface, instead of just adding the mode to the reaction method.

I guess either way this also means that we need to introduce the `get_mode()` method for reaction stores.

**Comment by JDGrimes** [comment]
Also, we've got to consider the implications of changing the reaction GUID format. Every reaction's GUID will be reset, which will mean that it will be as if the hook logs were wiped clean. Every reaction will behave as a totally new reaction. Thankfully, we're still in beta (or alpha or whatever), so I'm not really concerned about this. But doing something like this post-release will require a database upgrade.

**Comment by JDGrimes** [comment]
I think introducing a reaction store GUID would be for a separate ticket. I'm not going to base the reaction mode off of that, since it doesn't have to relate to the store by any constraint of the interface. In fact, we'll be introducing a `get_mode_slug()` method on the reaction interface as well, since reactions are by implication related to particular modes. And that is considered to be independent of the entire concept of reaction stores. Sort of.

**Comment by JDGrimes** [comment]
How we got into this position in the first place appears to be that we didn't properly update this when we severed reactors from reaction stores. Before that, the reaction store slug *was* the mode. But after that it became the "UI slug", which is more similar to what the reactor slug was before. Well, maybe not exactly, but the point is that it wasn't the mode anymore, and so the mode was no longer a part of the GUID.

https://github.com/WordPoints/hooks-api/commit/eec1cfd656a4e2b22fe416c3f7ffb62cc42a8649#diff-d346cbe81c1effa70116445f3a429058L63
https://github.com/WordPoints/hooks-api/commit/eec1cfd656a4e2b22fe416c3f7ffb62cc42a8649#diff-2ba7fbc625dd6e77f48a3bebc844cedfL50
